### ISSUE TITLE
Kn plugin default path and lookup order

### ISSUE THREAD

**Issue by maximilien** [knative/client#269]
## In what area(s)?

Kn Plugins

Classifications:

> /kind proposal

## Describe the feature:

In the current Kn Plugins #249 the `—plugin-dir` flag is used as a persistent flag to allow the override of plugins search path. The current search order is:

1. `—plugin-dir` if specified
2.  `~/.kn/config.yaml` if that has `pluginDir` specified in it
3. $PATH 

Also if 1 is specified, `kn` will modify the value in the config to reflect the value specified.

There are various discussion on whether we should change this order and whether we should specify a specific directly as the path for plugins and whether we should save the `—plugin-dir` after execution (success, failure, or both).

@rhuss suggested the following from comment in #249 

```text
I highly would prefer that `~/.kn/plugins/` is the default with either an option in the configuration to lookup in the path, too (e.g. lookupPath: true) or having path lookup as a fallback, with also looking up in ~/.kn/plugins. 

Looking up the whole path is an error-prone process (e.g. think about network mounted directories, bogus pathes added, the sheer amount of files to check (>2k for my machine). My suggestion for the lookup order would be

* Lookup ~/.kn/plugins (or the directory is given as config option pluginDir:, take if found (even shorter ones)
* Lookup in $PATH as fallback but only if the not pathLokup: false is set.
```

My views are the search speed is not a large concern for most users (otherwise they would have changed their PATH to allow faster lookup) and more importantly it’s a problem I have never seen myself on modern Linux kernel and filesystem. But I will concede that servers and VMs might have different characteristics. Also, users can always specify a specific directory for plugins. So from that perspective the `Kn plugins` #249 already implemented specific directory. 

*The real question is what default and search order is most sensible?*

**Comment by rhuss** [comment]
Speed is no concern, indeed.

It's about serving two valid use cases at the same time out-of-the-box without special configuration. 

I copied over my other attempt to clarify my motivation for this suggestion as stated in https://github.com/knative/client/pull/249#issuecomment-512206228

Maybe let me rephrase my suggested compromise for covering those two use cases *out of the box without any extra configuration* :

* **Use Case A**: I want to store my plugins anywhere in path with a `kn-` prefix and I don't care about any plugin directory (like `kubectl` does)
* **Use Case B**: I want to store my plugins in a dedicated directory, so that I know where I can put it in and where kn could automatically put it when downloaded from a remote repository** (like [krew](https://github.com/kubernetes-sigs/krew))

The compromise algorithm for looking up a plugin looks like this:

1. Check for the plugin in `~/.kn/plugins` first. If this directory exists and plugin is in there, use that.
2. If not found in step 1, check the whole $PATH for the plugin name, prefixed with `kn-`. Pick it if found.
3. Return an error if not found.

As configuration variable use a `lookupPluginsInPath`. If set to `false` then skip step 2. The default value is `true`. Should be a config variable only (no command line flag)

Optionally, introduce a configuration option `pluginDirectory`, which would override the default `~/.kn/plugins`. But we could as well keep it hardcoded to `~/.kn /plugins` as well.

-------------------

This would support both use cases out of the box **without any configuration**. 

* **A Users** who don't want a plugin directory, just don't create it or leave it empty. They put all their plugins somewhere into the path
* **B Users** put all the plugins into plugin directory. This works ootb, but they might consider to add `lookupPluginsInPath` to false to be sure to not pick up a plugin accidentally.

--------------------

Would this something we could agree on ?

**Comment by maximilien** [comment]
At this point I just want us to have the plugin support added and get this in front of users. There is a lot more to do to get plugins usable beyond search and plugin location. So want to get to that.

W.r.t. `pluginDir` I think @sixolet wanted to specify specific directory, especially for the server-side plugins she wanted to support. Let's see if she can live with `~/.kn/plugins` or want to allow any directory to be specified.

---
I am somewhat ambivalent. This proposal essentially flips the order of plugin lookup and makes the `$PATH` search optional. So relatively easily doable. I'll wait for about 24 hours to see if others want to chime, otherwise, I'll change to it. If something comes up, I'll ping you on Slack and here.

**Comment by maximilien** [comment]
I think this is done now with #249 merged. I will close this. @rhuss please re-open if anything missing.
